File: jupyter-gateway/src/main/java/com/example/jupytergateway/dto/response/R.java
package com.example.jupytergateway.dto.response;

import lombok.*;
import lombok.experimental.Accessors;

import java.io.Serial;
import java.io.Serializable;

@Builder
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
public class R<T> implements Serializable {
    @Serial
    private static final long serialVersionUID = 1L;

    public static final int CODE_SUCCESS = 0;
    public static final int CODE_ERROR = -500;
    public static final int BAD_REQUEST = -400;

    @Getter
    @Setter
    private int code;

    @Getter
    @Setter
    private String msg;

    @Getter
    @Setter
    private T data;

    public static R<Object> ok() {
        return new R<>(CODE_SUCCESS, "ok", null);
    }

    public static <T> R<T> ok(T data) {
        return new R<>(CODE_SUCCESS, "ok", data);
    }

    public static <T> R<T> ok(T data, String msg) {
        return new R<>(CODE_SUCCESS, msg, data);
    }

    public static R<Object> error() {
        return new R<>(CODE_ERROR, "error", null);
    }

    public static R<Object> error(String msg) {
        return new R<>(CODE_ERROR, msg, null);
    }

    public static R<Object> badRequest(String msg) {
        return new R<>(BAD_REQUEST, msg, null);
    }

    public String toString() {
        return "{\"code\": " + this.getCode() + ", \"msg\": " + this.transValue(this.getMsg()) + ", \"data\": " + this.transValue(this.getData()) + "}";
    }

    private String transValue(Object value) {
        return value instanceof String ? "\"" + value + "\"" : String.valueOf(value);
    }
}

File: api-factory/af-security-gateway/src/main/kotlin/org/inet/aet/afsecuritygateway/dto/response/R.kt
package org.inet.aet.afsecuritygateway.dto.response

import lombok.*
import lombok.experimental.Accessors
import java.io.Serial
import java.io.Serializable

@Builder
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
class R<T>(
    var code: Int,
    var msg: String?,
    var data: T?,
) : Serializable {

    private fun transValue(value: Any): String {
        return if (value is String) "\"" + value + "\"" else value.toString()
    }

    companion object {
        @Serial
        private val serialVersionUID = 1L

        const val CODE_SUCCESS: Int = 0
        const val CODE_ERROR: Int = -500
        const val BAD_REQUEST: Int = -400
        const val FORBIDDEN: Int = -403
        const val SERVICE_UNAVAILABLE: Int = -503
    }
}

fun <T> R_SUCCESS(data: T): R<T> {
    return R(R.CODE_SUCCESS, "success", data)
}


fun R_FORBIIDEN(msg: String): R<Any> {
    return R(R.FORBIDDEN, "fail", null)
}